File: software_architecture_model_python/src/software_architecture_model_python/schemas/registry.py
"""
Μηχανισμός φόρτωσης canonical αρχιτεκτονικών trees από το directory `model/`.

Το module παρέχει την κλάση SchemaRegistry, η οποία:
- εντοπίζει τα JSON αρχεία κάθε αρχιτεκτονικής (MVC, MVVM, DDD, Event‑Driven, Flow‑Based)
- τα φορτώνει σε δομές Python
- τα επιστρέφει στον ταξινομητή αρχιτεκτονικής (ArchitectureClassifier)

Χρησιμοποιείται ως κεντρικό registry για όλα τα canonical trees.
"""

import json
from pathlib import Path
from typing import Any, cast


class SchemaRegistry:
    """
    Φορτώνει όλα τα canonical architecture trees από το repo.
    """

    def __init__(self) -> None:
        self.base_path = Path(__file__).resolve().parent.parent.parent.parent / "model"

    def load_json(self, path: Path) -> dict[str, Any]:
        """
    Φορτώνει ένα JSON αρχείο και επιστρέφει το περιεχόμενό του ως dict.

    :param path: Το μονοπάτι του JSON αρχείου.
    :type path: Path
    :return: Τα δεδομένα του JSON ως λεξικό.
    :rtype: dict[str, Any]
    """
        data = json.load(path.open("r", encoding="utf-8"))
        return cast(dict[str, Any], data)

    def load_architecture(self, arch: str) -> dict[str, Any]:
        """
    Φορτώνει όλα τα JSON trees για μια συγκεκριμένη αρχιτεκτονική.

    :param arch: Το όνομα της αρχιτεκτονικής (π.χ. 'MVC').
    :type arch: str
    :return: Λεξικό με όλα τα trees της αρχιτεκτονικής.
    :rtype: dict[str, Any]
    """
        arch_path = self.base_path / arch
        trees: dict[str, Any] = {}

        for json_file in arch_path.glob("*.json"):
            trees[json_file.stem] = self.load_json(json_file)

        return trees

    def load_all(self) -> dict[str, dict[str, Any]]:
        """
    Φορτώνει όλες τις αρχιτεκτονικές και τα canonical trees τους.

    :return: Λεξικό με όλες τις αρχιτεκτονικές και τα trees τους.
    :rtype: dict[str, dict[str, Any]]
    """
        architectures = [
            "DDD",
            "Event-Driven",
            "Flow_Based_Architecture",
            "MVC",
            "MVVM",
        ]

        registry: dict[str, dict[str, Any]] = {}

        for arch in architectures:
            registry[arch] = self.load_architecture(arch)

        return registry

File: js_package/src/schemas/registry.ts
/**
 * Μηχανισμός φόρτωσης canonical αρχιτεκτονικών trees από το directory `model/`.
 *
 * Η κλάση SchemaRegistry:
 * - εντοπίζει τα JSON αρχεία κάθε αρχιτεκτονικής (MVC, MVVM, DDD, Event‑Driven, Flow‑Based)
 * - τα φορτώνει σε δομές JS/TS
 * - τα επιστρέφει στον ταξινομητή αρχιτεκτονικής
 *
 * Χρησιμοποιείται ως κεντρικό registry για όλα τα canonical trees.
 */
import { fileURLToPath } from "node:url";
import path from "path";
import { promises as fs } from "fs";

const __filename = fileURLToPath('import.meta.url');
const __dirname = path.dirname(__filename);

export class SchemaRegistry {
  private basePath: string;

  constructor() {
    this.basePath = path.resolve(__dirname, "../../../model");
  }

  async loadJson(filePath: string): Promise<Record<string, any>> {
    const raw = await fs.readFile(filePath, "utf-8");
    return JSON.parse(raw);
  }

  async loadArchitecture(arch: string): Promise<Record<string, any>> {
    const archPath = path.join(this.basePath, arch);
    const entries = await fs.readdir(archPath);

    const trees: Record<string, any> = {};

    for (const file of entries) {
      if (file.endsWith(".json")) {
        const fullPath = path.join(archPath, file);
        const stem = path.basename(file, ".json");
        trees[stem] = await this.loadJson(fullPath);
      }
    }

    return trees;
  }

  async loadAll(): Promise<Record<string, Record<string, any>>> {
    const architectures = [
      "DDD",
      "Event-Driven",
      "Flow_Based_Architecture",
      "MVC",
      "MVVM"
    ];

    const registry: Record<string, Record<string, any>> = {};

    for (const arch of architectures) {
      registry[arch] = await this.loadArchitecture(arch);
    }

    return registry;
  }
}
